### ISSUE TITLE
support the structure of arm32, arm64, and x86 at the same time?

### ISSUE THREAD

**Issue by xujingzhou** [bytedeco/javacpp#632]
Hi, What should I do to enable the generated jni. so library to support the structure of arm32, arm64, and x86 at the same time?

**Comment by saudet** [comment]
You mean using something like Apple's compiler? We can create platform properties based on the default ones available here:
https://github.com/bytedeco/javacpp/tree/master/src/main/resources/org/bytedeco/javacpp/properties
BTW, contributions are welcome, so please open a pull request to have your new properties included there! Thanks

**Comment by saudet** [comment]
I don't believe Android supports libraries that can support multiple architectures. If you're not asking about libraries that can support multiple architectures, you'll need to clarify your question to get a useful answer.

**Comment by xujingzhou** [comment]
Sorry, I am trying to generate an.aar file for android with internal contents ("armeabi-v7a", "ar64-v8a ", "x86_64").

**Comment by xujingzhou** [comment]
Thanks.  I'm going to try an example as below:
https://github.com/bytedeco/sample-projects/blob/master/javacpp-android-gradle/app/build.gradle

**Comment by saudet** [comment]
Right, you probably want to use Gradle JavaCPP for that: https://github.com/bytedeco/gradle-javacpp

**Comment by xujingzhou** [comment]
The following error was encountered when compiling the following example.

sample as below:
https://github.com/bytedeco/sample-projects/blob/master/javacpp-android-gradle/

Error as below:
++javacpp file path:/root/.gradle/caches/modules-2/files-2.1/org.bytedeco/javacpp/1.5.8/a28ca7c27abaae8054060b963fbd667b4be72677/javacpp-1.5.8.jar
/home/android-sdk/ndk/24.0.8215888
/root/.gradle/caches/modules-2/files-2.1/org.bytedeco/javacpp/1.5.8/a28ca7c27abaae8054060b963fbd667b4be72677/javacpp-1.5.8.jar

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Cannot create variant 'android-manifest-metadata' after dependency configuration ':app:debugApiElements' has been resolved

**Comment by saudet** [comment]
Please try to use Gradle JavaCPP instead: https://github.com/bytedeco/gradle-javacpp

**Comment by xujingzhou** [comment]
The follow sample has been fixed.

Sample: https://github.com/bytedeco/sample-projects/blob/master/javacpp-android-gradle/

grade: 6.9.3
com.android.tools.build:gradle:4.2.0

Fixed build.gradle:
apply plugin: 'com.android.application'
def ndkPath

android {
    compileSdkVersion 27
    buildToolsVersion "28.0.3"
    defaultConfig {
        applicationId "idl.baidu.com.javacppgradle"
        minSdkVersion 22
        targetSdkVersion 27
        versionCode 1
        versionName "1.0"
        externalNativeBuild {
            cmake {
                cppFlags "-std=c++11 -frtti -fexceptions"
            }
        }
        ndk {
            // Specifies the ABI configurations of your native
            // libraries Gradle should build and package with your APK.
            abiFilters 'x86', 'x86_64', 'armeabi-v7a','arm64-v8a'
        }
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }

    lintOptions {
       abortOnError false
    }

    externalNativeBuild {
        android {
            externalNativeBuild {
                ndkPath = '/home/android-sdk/ndk/21.4.7075529'
            }
        }
        cmake {
            path "CMakeLists.txt"
        }
    }
    sourceSets.main {
        jniLibs.srcDir 'src/main/libs'
        jni.srcDirs = [] //disable automatic ndk-build call
    }
}

android.applicationVariants.all { variant ->
    println("-----------------------")
    println 'variant.name: ' + variant.name
    def projectDir = project.projectDir.absolutePath
    def classesDir = variant.javaCompileProvider.get().destinationDir
    def jniDir = file("$projectDir/src/main/cpp/").absolutePath
    println("projectDir: " + projectDir)
    println("classesDir: " + classesDir)
    println("jniDir: " + jniDir)
    println("ndkPath: " + ndkPath)
    // NOTE: The functionlity of javacpp is create jni cpp file from java code.++++++++++++++++
    // javacpp workflow:
    // 1.generate jni cpp file by javacpp-java code file.
    // 2.nkd build with android.mk
    variant.javaCompileProvider.get().doLast {
        println '===javacpp start==='
        javaexec {
            main = 'org.bytedeco.javacpp.tools.Builder'
            //NOTE: jar version must be same with the dependency jar.
            classpath = files("$projectDir/javacpp-platform-1.5.8-bin/javacpp.jar")
            args '-cp', "$classesDir",
                    '-properties', 'android-arm',
                    '-Dplatform.root=' + ndkPath,
                    '-Dplatform.compiler=' + ndkPath + '/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++',
                    //NOTE: include path will auto work by javacpp.jar 1.1 version.(see Android.mk)
//                    "-Dplatform.includepath=" + includePath+"/include:" +includePath+"/libs/armeabi/include",
                    "-nocompile", "idl.baidu.cpp.*",
                    "-d", "${jniDir}" // jni file destination folder.
        }
        println '===javacpp done==='
    }
}

dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation 'com.android.support:appcompat-v7:27.1.1'
    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
    implementation 'org.bytedeco:javacpp:1.5.8'
}

**Comment by saudet** [comment]
Great! Please open a pull request: https://github.com/bytedeco/sample-projects/pulls

**Comment by lancewoo** [comment]
> The follow sample has been fixed.
> 
> Sample: https://github.com/bytedeco/sample-projects/blob/master/javacpp-android-gradle/
> 
> grade: 6.9.3 com.android.tools.build:gradle:4.2.0
> 
> Fixed build.gradle:
> 
> ```groovy
> apply plugin: 'com.android.application'
> def ndkPath
> 
> android {
>     compileSdkVersion 27
>     buildToolsVersion "28.0.3"
>     defaultConfig {
>         applicationId "idl.baidu.com.javacppgradle"
>         minSdkVersion 22
>         targetSdkVersion 27
>         versionCode 1
>         versionName "1.0"
>         externalNativeBuild {
>             cmake {
>                 cppFlags "-std=c++11 -frtti -fexceptions"
>             }
>         }
>         ndk {
>             // Specifies the ABI configurations of your native
>             // libraries Gradle should build and package with your APK.
>             abiFilters 'x86', 'x86_64', 'armeabi-v7a','arm64-v8a'
>         }
>     }
>     buildTypes {
>         release {
>             minifyEnabled false
>             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
>         }
>     }
> 
>     lintOptions {
>        abortOnError false
>     }
> 
>     externalNativeBuild {
>         android {
>             externalNativeBuild {
>                 ndkPath = '/home/android-sdk/ndk/21.4.7075529'
>             }
>         }
>         cmake {
>             path "CMakeLists.txt"
>         }
>     }
>     sourceSets.main {
>         jniLibs.srcDir 'src/main/libs'
>         jni.srcDirs = [] //disable automatic ndk-build call
>     }
> }
> 
> android.applicationVariants.all { variant ->
>     println("-----------------------")
>     println 'variant.name: ' + variant.name
>     def projectDir = project.projectDir.absolutePath
>     def classesDir = variant.javaCompileProvider.get().destinationDir
>     def jniDir = file("$projectDir/src/main/cpp/").absolutePath
>     println("projectDir: " + projectDir)
>     println("classesDir: " + classesDir)
>     println("jniDir: " + jniDir)
>     println("ndkPath: " + ndkPath)
>     // NOTE: The functionlity of javacpp is create jni cpp file from java code.++++++++++++++++
>     // javacpp workflow:
>     // 1.generate jni cpp file by javacpp-java code file.
>     // 2.nkd build with android.mk
>     variant.javaCompileProvider.get().doLast {
>         println '===javacpp start==='
>         javaexec {
>             main = 'org.bytedeco.javacpp.tools.Builder'
>             //NOTE: jar version must be same with the dependency jar.
>             classpath = files("$projectDir/javacpp-platform-1.5.8-bin/javacpp.jar")
>             args '-cp', "$classesDir",
>                     '-properties', 'android-arm',
>                     '-Dplatform.root=' + ndkPath,
>                     '-Dplatform.compiler=' + ndkPath + '/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-g++',
>                     //NOTE: include path will auto work by javacpp.jar 1.1 version.(see Android.mk)
> //                    "-Dplatform.includepath=" + includePath+"/include:" +includePath+"/libs/armeabi/include",
>                     "-nocompile", "idl.baidu.cpp.*",
>                     "-d", "${jniDir}" // jni file destination folder.
>         }
>         println '===javacpp done==='
>     }
> }
> 
> dependencies {
>     implementation fileTree(dir: 'libs', include: ['*.jar'])
>     implementation 'com.android.support:appcompat-v7:27.1.1'
>     implementation 'com.android.support.constraint:constraint-layout:1.1.3'
>     implementation 'org.bytedeco:javacpp:1.5.8'
> }
> ```

You may check this project [gradle-javacpp-android](https://github.com/bytedeco/sample-projects/tree/master/gradle-javacpp-android) or [this one](https://github.com/bytedeco/sample-projects/tree/master/laslib-android)
